src/storage/files.c

```diff
@@ -339,11 +344,145 @@ static int password_callback(char* buf, int bufsiz, int verify, void* cb_tmp)
         {
             res = bufsiz;
         }
-        memcpy(buf, password, res); /* copy password and length(res) into buf */
+        memcpy(buf, password, (size_t)res); /* copy password and length(res) into buf */
     }
     return res; /* the size */
 }
 
+
+static int get_algo_nid(const X509_ALGOR *alg)
+{
+    int pbenid, aparamtype;
+    const ASN1_OBJECT *aoid;
+    const void *aparam;
+
+    X509_ALGOR_get0(&aoid, &aparamtype, &aparam, alg);
+
+    pbenid = OBJ_obj2nid(aoid);
+
+    if(pbenid == NID_pbes2)
+    {
+        PBE2PARAM *pbe2 = 0;
+        int encnid;
+        if(aparamtype == V_ASN1_SEQUENCE)
+        {
+            pbe2 = ASN1_item_unpack(aparam, ASN1_ITEM_rptr(PBE2PARAM));
+        }
+        if(pbe2 == 0)
+        {
+            BIO_puts(bio_err, ", <unsupported parameters>");
+            return NID_undef;
+        }
+        X509_ALGOR_get0(&aoid, &aparamtype, &aparam, pbe2->keyfunc);
+        pbenid = OBJ_obj2nid(aoid);
+        X509_ALGOR_get0(&aoid, 0, 0, pbe2->encryption);
+        encnid = OBJ_obj2nid(aoid);
+        PBE2PARAM_free(pbe2);
+        return encnid;
+    }
+
+    return NID_undef;
+}
+
+
+static int bag_get_pkey_algo(const PKCS12_SAFEBAG *bag, const char *pass, int passlen)
```

Comment:
`pass` and `passlen` are unused, so remove them from the function parameter list.